### DISCUSSION TITLE
Fall 2024 Breakages

### DISCUSSION THREAD

**Discussion by magnus-madsen** [flix/flix#9089] [General]
We are planning to rework and remove some features which will cause breakages (sorry...)

Features to be reworked or removed:

- We begin to drop old Java interop support (`import get ...` etc.)
- We drop `!` from functions, e.g. `MutMap.put!` with become just `MutMap.put`.
- We will start writing `Heap[r]` instead of just `r` for region effects.
- We will drop `File` and `Files` which will be replaced by algebraic effects and/or just plain old Java (via interop).
- We will likely replace `Enviroment` by an algebraic effect.
- We will drop the short-hand effect syntax `{ef, IO}` which will now have to be written `ef + IO`.
- Casts will be slightly reworked (again, again, again, sigh.)

We will also introduce some new features that cause breakages:

- We will begin introducing primitive effects (formerly foundational effects, formerly base effects).
- These are effects that will decorate the Java library. 
  - For example, using FileReader will now have both the `FsRead` and `IO` effects. 

While breakages are painful, I believe all of these changes are for the better.

**Comment by magnus-madsen** [comment]
@JonathanStarup @mlutze @jaschdoc @stephentetley CC

**Comment by magnus-madsen** [comment]
While `File`, `Files` and `Environment` are to be removed, it would be easy for anyone to extract them into its own package, if desired. They are just thin wrappers with the `IO` effect. In the near future `FileRead` and `FileWrite` AE will provide the same (but of course better) functionality (see e.g. #9065)

**Comment by JonathanStarup** [comment]
so effect syntax doesnt include {}? or {Crash, IO}

**Reply by magnus-madsen** [reply]
No, you still have that.

**Reply by magnus-madsen** [reply]
The point was to remove effect variables as members of a set literal. That's what you wanted right?
